### ISSUE TITLE
Some things about the PushEncryptionKeys p256dh and auth are not clear enough

### ISSUE THREAD

**Issue by PascalReintjens** [w3c/push-api#212]
In the working draft 15 is written: 

> Keys named p256dh and auth must be supported.

Does that mean that those keys must exist and contain a value, is it enough that they exist and contain null or does that mean (in context of what comes before that sentence) that if getKey() is supported, only then you must be able to get auth and p256dh?

Secondly, what is not clear to me is, whether it is possible that auth and p256dh can change after the subscription:
Case 1: If the values were null, is it possible that a value gets added without a new subscription - For example: a browser wasn't able to generate these values and was now updated to a version which can
Case 2: A browser had values, is it possible that these get deleted or changed without a new subscription?
So basically the question is can I assume that in any case the subscription settings stay exactly the same until the endpoint is invalidated?

In the editor draft from 18 August 2016 is written: 

> Once set, options for a push subscription cannot change.

But that is in a part which focuses on the  PushSubscriptionOptions userVisibleOnly and applicationServerKey. And if that also contains p256 and auth, then the question is again: does "set" mean filled with a value or does null qualify too, that the "value" cannot be changed?

Thanks

**Comment by beverloo** [comment]
The `getKey()` method is a required for an implementation of the Push API, so the `auth` and `p256dh` fields must be supported too. Would updating the cited sentence to the following clarify the statement?

> Keys named `p256dh` and `auth` must be supported, and their values must correspond to those necessary for the user agent to decrypt received push messages in accordance with [WEBPUSH-ENCRYPTION].

In Chrome, we introduced support for encrypted payloads in version 50 by adding the `getKey()` method, updating the `PushSubscription` serializer and lazily creating keys for subscriptions created prior to that release. I guess new, partial implementations could do something similar. This would significantly impact their interoperability story, so it's undesirable at best.

The keys are associated with the subscription and should not change during its lifetime. If the user agent has to update them, the `pushsubscriptionchange` event should be used to communicate this to the developer. It'll be considered an entirely new subscription at that point.

I think step 15 of the subscribe() algorithm details this. It's the only place where a key pair is created.

**Comment by PascalReintjens** [comment]
> The getKey() method is a required for an implementation of the Push API, so the auth and p256dh fields must be supported too. Would updating the cited sentence to the following clarify the statement?
> 
> > Keys named p256dh and auth must be supported, and their values must correspond to those necessary for the user agent to decrypt received push messages in accordance with [WEBPUSH-ENCRYPTION].

Yes, I think that would make it clearer

---

> In Chrome, we introduced support for encrypted payloads in version 50 by adding the getKey() method, updating the PushSubscription serializer and lazily creating keys for subscriptions created prior to that release. I guess new, partial implementations could do something similar. This would significantly impact their interoperability story, so it's undesirable at best.

In some instances, this could be very problematic – for example: In one of my planned use cases, website visitors don't have a local account, if they want notifications they just press a toggle button on the site. So, then my server gets the endpoint, the auth-key-value/null and p256dh-key-value/null and that's it. To later set/update the settings of an existing notification or remove notifications, the browser now gives the endpoint to my webpage as “authentication”. That works, but is not very secure. For example, if there is a small push service which distributes very short endpoints it could be possible to brute force endpoints until a valid endpoint is found (which would be slowed down by brute force protection and you only would get access to a random endpoint-“account” but still with enough effort, it wouldn’t be impossible). Additionally, the more endpoints register, the higher the chance to find a valid endpoint per brute force attempt. To make this safer, I’d like to hash the auth-key and send this as second authentication characteristic, so that an attacker would have to exactly guess an endpoint + auth-key-hash combination. If there are then more endpoints, the auth-key-hash would still provide the same security (because different endpoints could theoretically have the same auth-key-hash). But this only works if I can be totally sure, that an endpoint without keys would always trigger pushsubscriptionchange as soon as endpoint-keys are added and that existing keys are not changed without calling pushsubscriptionchange. 

---

> The keys are associated with the subscription and should not change during its lifetime. If the user agent has to update them, the pushsubscriptionchange event should be used to communicate this to the developer. It'll be considered an entirely new subscription at that point.
> I think step 15 of the subscribe() algorithm details this. It's the only place where a key pair is created.

Step 15 in the editor draft from 18. August 2016 is:

> Generate a new P-256 ECDH key pair. Store the private key in an internal slot on subscription; this value must not be made available to applications. The public key is also stored in an internal slot and can be retrieved by calling the getKey method of the PushSubscription with an argument of p256dh.

So, while you could assume that the keys should not change, it is nowhere written down, which doesn't make it a good idea or safe to rely on.

How about adding something like?:

> If Keys named p256dh and auth of an push subscription are to be changed or added after the subscribe method is finished, this MUST result in calling pushsubscriptionchange

or if you don't want to let this result in a completely new subscription, you could add a small and simple function like pushsubscriptionkeyupdate, which basically would get the "old" subscription and a new one with updated keys.

Thanks for your support :)

**Comment by ghost** [comment]
> But this only works if I can be totally sure, that an endpoint without keys would always trigger pushsubscriptionchange as soon as endpoint-keys are added and that existing keys are not changed without calling pushsubscriptionchange. 

Maybe it's also worth calling out as an example in 11.2: "For example, the push service MAY set an expiration time, or the user agent MAY rotate the push message encryption keys."

You're completely correct: if the keys change, the service worker and app server need to know. Otherwise, any messages sent to that subscription using the old keys will be dropped, because the browser will fail to decrypt them with the new keys.

**Comment by PascalReintjens** [comment]
Sorry for the late response, I was sick the last few days :(

> Maybe it's also worth calling out as an example in 11.2: "For example, the push service MAY set an expiration time, or the user agent MAY rotate the push message encryption keys."

Yes, that would improve the situation by at least suggesting the right action, although that still leaves it at a state where the behavior is completely optional

> You're completely correct: if the keys change, the service worker and app server need to know. Otherwise, any messages sent to that subscription using the old keys will be dropped, because the browser will fail to decrypt them with the new keys.

That is the point - the key changes need to be communicated so that message sending still works. Even adding keys without notifying service worker and app server has no point, since they then are still not able to use them. Because changing without notifying is not a good idea in any case, why not make the notifying a must so that the push api gets more reliable? 

Just to get the idea - my rough example, I'm sure you can formulate that way better than me:
`If Keys named p256dh and auth of an push subscription are changed or added after the subscribe method is finished, this MUST be communicated through pushsubscriptionchange`

Thanks :)

**Comment by ghost** [comment]
Sorry for asking question out of context 
A thing is bugging me for some days that how would I send notification to a particular user whose I have auth, p256dh and endpoint.
I have many of these for different user, which of these will help me to distinguish one user from other.
Please I really stuck over this though from many days.

**Comment by nxthor** [comment]
late to the game, but maybe of help regarding that last question.

When you ask for the subscription on the client side and send the subscription information to the server, you can make sure to have a login or some other identifier to distinguish your users and add that data. Otherwise all the subscriptions are basically anonymous if you leave out the user agent. In my case I only ask for subscription data, when the user is logged in. Next I can save the subscription with the user in reference on the server side. When sending something to certain users (notifications on personal events in the application) I have one or more subscriptions to send to. If any error occurs, that leads to a retracted/general broken subscription, I delete the subscription on my side (or flag it). Since I only send to certain users, I have no use case like "send to everyone", that may differ from scenarios like a newspaper urging everyone to read a new article without bothering who it is, as long as there is a valid subscription in storage.
